### PR TITLE
Doc update for SC17

### PR DESCRIPTION
**Pull request by krafczyk**
Hey folks,

I've updated a few sections of the documentation for SC17. I've rearranged one section, and added sections on topics which aren't much discussed elsewhere. These are all things I had trouble figuring out on my own, and I think the documentation would benefit from having them in there.

Topics include:
- Better enumeration of a spec objects abilities
- Examples of packages to use for inspiration
- Another mention of using the `run_before` and `run_after` decorators
- How to get configuration variables within spack
- How to change or add configuration variables to spack
- How to execute command line programs
- A small mention about how hashes are calculated.

Give me feedback on formatting and content.

### CODE REVIEW

**Comment by krafczyk** [comment]
Incidentally, this also _kinda_ solves #2016.

File: lib/spack/docs/packaging_guide.rst [hunk 1]
```diff
@@ -2399,6 +2413,56 @@ the install locations of dependencies, or when you need to do something differen
 depending on the version, compiler, dependencies, etc. that your package is
 built with.  These parameters give you access to this type of information.
 
+^^^^^^^^^^^^^^^^^
+Adding new phases
+^^^^^^^^^^^^^^^^^
+
+Phases can be added before or after existing phases by using the ``run_before``
```
Comment (alalazo) [inline lib/spack/docs/packaging_guide.rst]: That's not really correct. The *phases* are specified in the base build-system class, but you can run arbitrary code using the `@run_before(phase)` and `@run_after(phase)` decorator.

Comment (alalazo) [reply lib/spack/docs/packaging_guide.rst]: For instance, every `CMakePackage` has 3 phases:

https://github.com/LLNL/spack/blob/9aafe21b6328d5944469b7a0933e5a446133f876/lib/spack/spack/build_systems/cmake.py#L65-L66

but any package can attach code in between.

Comment (krafczyk) [reply lib/spack/docs/packaging_guide.rst]: I've updated the wording. Have a look!

**Comment by adamjstewart** [comment]
It looks like this PR is quite out-of-date. Much of the docs added here have since been added to the Build System docs or the Configuration docs. I'm going to close this, but feel free to reopen if you want to update this.

**Pull request closed without merging**